python/ray/air/result.py

```diff
@@ -21,12 +23,21 @@ class Result:
     Args:
         metrics: The final metrics as reported by an Trainable.
         checkpoint: The final checkpoint of the Trainable.
+        best_checkpoint: The best checkpoint of the Trainable, as
```

Comment:
do we have concerns over conflicting messages? Or do we have an idea to make it more clear?
for example,
how do we reconcile with the fact that Tune scheduler pretty much only looks at the last result of each trial and make a decision on whether to proceed or not. In this context, `mode` and `metrics` are used in a horizontal context.
on the other hand, `best_checkpoint` here is looking at things vertically.
But somehow, `best_checkpoint` is also using TuneConfig (from where tune `mode` and `metrics` are extracted).
My worry is users may not get the exact right mental picture.

Reply:
Yeah it's tricky. In general, this is how I imagine the user would want things to happen:
1. If I am tuning, then I am maximizing/minimizing a metric, as set in `TuneConfig`. In most cases, the best checkpoint is the checkpoint that does that
2. I want to be able to have a way to override that default behavior, however, if need be
3. If I am training, then I don't have access to `TuneConfig`, so I need a way to set this behavior in some other way

Following that, it made sense for me to include `TuneConfig` here as well (if we are maximing accuracy then the best checkpoint will obviously be the one that maximizes accuracy). I think it would be fine to make this dependent only on `RunConfig.checkpointing`, but that may lead to some confusion when users do not set it

Reply:
Yeah I think the CUJ you have in mind makes sense. 
Just want to explore any housekeeping that can make the message more clear. For example (very naively, looking for something better...)
Should we have a MetricConfig thing in air/config.py, and explicitly document that this is used in 
1. horizontally across tune trials
2. vertically within each trial when keeping and retrieving best checkpoint

Obviously, I don't like that there is one more layer into configuring things...

Reply:
I think that may be a bit too much, yeah. I feel like improving this docstring should be enough (maybe have a numbered list)

Reply:
I personally feel that the behavior described in this docstring is pretty confusing. Do we have to expose all these details to users?

Reply:
@xwjiang2010 I updated the docstring, ptal

Reply:
@amogkam ideally, for the user, it will "just work". What do you find confusing here? Can you elaborate?